File: apps/marquee/quorum-sense/ios/CoreBridge/RefineServiceLlm.swift
import Foundation

/// Cloud-fallback LLM backend (M6 compute placement). Implements the generated
/// UniFFI `LlmBackend` callback: the Rust refinement loop calls `complete`, and
/// this POSTs the prompt to the local Quorum refine-service, which holds the API
/// keys (so no key ever reaches the device). Returns `nil` on any failure — the
/// Rust refiner then falls back to its deterministic heuristics, so a draft is
/// always produced.
///
/// Blocking by design: `complete` is called synchronously from Rust, which runs
/// off the main thread behind the `QuorumCoreBridgeFFI` actor (ADR 0003), so
/// waiting on the request here never freezes the UI. iOS exempts loopback
/// (127.0.0.1) from App Transport Security, so cleartext to the local service
/// needs no Info.plist exception.
public final class RefineServiceLlm: LlmBackend, @unchecked Sendable {
    // iOS Simulator reaches the Mac's localhost directly. (A device build would
    // point this at the GC-Secrets backend instead.)
    private let endpoint: URL
    private let session: URLSession

    public init(endpointString: String = "http://127.0.0.1:8765/complete") {
        self.endpoint =
            URL(string: endpointString) ?? URL(string: "http://127.0.0.1:8765/complete")!
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 20
        config.waitsForConnectivity = false
        self.session = URLSession(configuration: config)
    }

    public func complete(prompt: String) -> String? {
        guard let body = try? JSONSerialization.data(withJSONObject: ["prompt": prompt]) else {
            return nil
        }
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = body

        let semaphore = DispatchSemaphore(value: 0)
        var result: String?
        let task = session.dataTask(with: request) { data, _, _ in
            defer { semaphore.signal() }
            guard
                let data,
                let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                let text = json["text"] as? String,
                !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            else { return }
            result = text
        }
        task.resume()
        semaphore.wait()
        return result
    }
}

File: apps/marquee/quorum-sense/android/app/src/main/java/se/reflective/quorum/corebridge/RefineServiceLlm.kt
package se.reflective.quorum.corebridge

import java.net.HttpURLConnection
import java.net.URL
import org.json.JSONObject
import uniffi.quorum_ffi.LlmBackend

/**
 * Cloud-fallback LLM backend (M6 compute placement). Implements the generated
 * UniFFI [LlmBackend] callback: the Rust refinement loop calls [complete], and
 * this POSTs the prompt to the local Quorum refine-service, which holds the API
 * keys (so no key ever reaches the device). Returns `null` on any failure — the
 * Rust refiner then falls back to its deterministic heuristics, so a draft is
 * always produced.
 *
 * Blocking by design: [complete] is called synchronously from Rust, which runs
 * on `Dispatchers.Default` off the main thread, so waiting on the request here
 * never freezes the UI. `10.0.2.2` is the host loopback as seen from the Android
 * emulator; cleartext to it is allowed via `usesCleartextTraffic` in the debug
 * manifest. A device build would point this at the GC-Secrets backend instead.
 */
class RefineServiceLlm(
    private val endpoint: String = "http://10.0.2.2:8765/complete",
) : LlmBackend {
    override fun `complete`(`prompt`: String): String? {
        return try {
            val connection = (URL(endpoint).openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                connectTimeout = 5_000
                readTimeout = 20_000
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
            }
            connection.outputStream.use { out ->
                out.write(JSONObject().put("prompt", prompt).toString().toByteArray())
            }
            val text =
                if (connection.responseCode == 200) {
                    val body = connection.inputStream.bufferedReader().use { it.readText() }
                    JSONObject(body).optString("text", "")
                } else {
                    ""
                }
            connection.disconnect()
            text.ifBlank { null }
        } catch (e: Exception) {
            // Any network/parse failure -> null -> Rust heuristic fallback.
            null
        }
    }
}
